### DISCUSSION TITLE
1.7 递归函数

### DISCUSSION THREAD

**Discussion by giscus** [csfive/composing-programs-zh#103] [Announcements]
# 1.7 递归函数

SICP Python 中文翻译

https://composingprograms.netlify.app/1/7

**Comment by kirkzhangtech** [comment]
真的会在实际的应用中使用递归吗？

**Reply by liucheng532** [reply]
我也是上完后收获很大，改变了思考问题解决问题的思路，思维模式的改变有用

**Comment by Life-s-A-Struggle** [comment]
不太能理解分割数的算法T-T

**Reply by ciron-nine** [reply]
看成一整个集合，大集合由两个小集合（包含最大数、不包含最大数）构成，这两个小集合互斥，即两个小集合元素个数之和等于大集合元素个数之和。然后就是表达问题，求包含最大数的集合元素个数问题可以表达为：求该数减去一个最大数（最大数不变）集合元素个数问题，另一个不包含最大数的问题则表达为求该数（最大数减一）的问题。这两个问题再分别继续分割，直至为0,证明找到一种排列，小于0,则是证明该排列错误。

**Reply by wangding** [reply]
跟换硬币的问题思路类似

**Comment by CharlieCRX** [comment]
# 简单阅读1.7的要点

首先一句话总结本节内容：用递归调用来理解“函数抽象”的概念。

所以在这里`递归`是工具，是为了方便更好地理解函数抽象概念的。这里面涉及到的函数抽象概念就是递归的核心问题：如何将原问题分解为同类的子问题而解决。

想要通畅阅读本节，你是需要一定的背景知识的。背景知识如下（预计背景知识准备时间：0.5-1h）：

1. [递归是什么？](https://algo.itcharge.cn/09.Algorithm-Base/02.Recursive-Algorithm/01.Recursive-Algorithm/#_1-%E9%80%92%E5%BD%92%E7%AE%80%E4%BB%8B)（思考：递归分为递推和回归，这两个过程是按照什么原则进行的？）
2. [如何实现递归？](https://algo.itcharge.cn/09.Algorithm-Base/02.Recursive-Algorithm/01.Recursive-Algorithm/#_3-%E9%80%92%E5%BD%92%E4%B8%89%E6%AD%A5%E8%B5%B0)
3. [图形化执行递归函数](https://pythontutor.com/cp/composingprograms.html#mode=edit)：为了更好地理解递归的过程，将1.7材料中较为简单的递归打印函数：`cascade(n)`代码粘贴进去，走一遍看看效果
4. [斐波那契数列的递归应用](https://algo.itcharge.cn/09.Algorithm-Base/02.Recursive-Algorithm/01.Recursive-Algorithm/#_5-1-%E6%96%90%E6%B3%A2%E9%82%A3%E5%A5%91%E6%95%B0)
5. [什么是`分割数`](https://zh.wikipedia.org/wiki/%E6%95%B4%E6%95%B8%E5%88%86%E6%8B%86)？（1.7节的材料最后用分割数作为检验递归调用是否理解的范例）

带着一定的问题将上面涉及的背景知识看完，然后再通篇阅读1.7，就能明白这节讲什么了。

**Reply by wmjim** [reply]
这个补充的背景知识真好

**Reply by wangding** [reply]
递归的分类一，按递归是否直接调用自己，分为：
- 直接递归
- 间接递归（这里的互递归）

递归的分类二，按递归函数中调用自己的次数，分为：
- 线性递归
- 两路递归（这里的二叉树形递归）
- 多路递归
